Add node performance measuring guide

timstclair opened the pull request:
Add a development guide for measuring performance of node components.
The purpose of this guide is threefold:
1. Document the nuances of measuring kubelet performance so we don't
   forget or need to reinvent the wheel.
2. Make it easier for new contributors to analyze performance.
3. Share tips and tricks that current team members might not be aware
   of.

This document is not complete (see TODO at the bottom), but hopefully it
can grow as we develop more tools for performance testing.

/cc @kubernetes/goog-node

yujuhong commented on docs/devel/node-performance-testing.md:
```diff
@@ -0,0 +1,146 @@
+Run the test with `go run hack/e2e.go -v -test
+--test_args="--ginkgo.focus=resource\susage\stracking`. You may also with to customise the number of
+pods or other elements of the test (remember to rerun `make WHAT=test/e2e/e2e.test` after you do).
+
```
Maybe add a section about prometheus or just have a link pointing to a doc? I couldn't find any useful esiting doc other than https://github.com/kubernetes/kubernetes/blob/master/docs/devel/instrumentation.md though

timstclair replied:
I don't have any experience using prometheus, so I left this in the TODO section.

yujuhong commented on docs/devel/node-performance-testing.md:
````diff
@@ -0,0 +1,146 @@
+$ curl -G "http://localhost:8001/api/v1/proxy/nodes/${NODE}:10250/debug/pprof/profile?seconds=${DURATION_SECONDS}" > $OUTPUT
+$ KUBELET_BIN=_output/dockerized/bin/linux/amd64/kubelet
+$ go tool pprof -web $KUBELET_BIN $OUTPUT
+```
````
Add /pprof/heap as well?

timstclair replied:
Done.

yujuhong commented on docs/devel/node-performance-testing.md:
```diff
@@ -0,0 +1,146 @@
+#### Disabling Add-ons
+
+Disabling addons is simple. Just ssh into the Kubernetes master and move or delete the addon from
+`/etc/kubernetes/addons/`.
```
A link to the add-ons may be useful in case this document becomes outdated: https://github.com/kubernetes/kubernetes/tree/master/cluster/addons

timstclair replied:
Done.

Random-Liu commented:
Cool idea! Thanks! :)

timstclair commented:
Addressed feedback. Thanks!

vishh commented on docs/devel/node-performance-testing.md:
```diff
@@ -0,0 +1,152 @@
+
+#### Disabling Add-ons
+
+Disabling addons is simple. Just ssh into the Kubernetes master and move or delete the addon from
```
I'd say `move` to be able to restore the cluster state.

timstclair replied:
Done.

vishh commented on docs/devel/node-performance-testing.md:
```diff
@@ -0,0 +1,152 @@
+Performance will vary a lot between a node with 0 pods and a node with 100 pods. In many cases
+you'll want to make measurements with several different amounts of pods. Scaling a replication
+controller makes this easy, just make sure the system reaches a steady-state before starting the
+measurement. E.g. `kubectl scale replicationcontroller pause --replicas=100`
```
This statement assumes a single node cluster right?

timstclair replied:
Done.

vishh commented on docs/devel/node-performance-testing.md:
```diff
@@ -0,0 +1,152 @@
+- Expand cluster set-up section
+- (vishh) Measuring disk usage
+- (yujuhong) Measuring memory usage
+- Add section on monitoring with prometheus
```
Alternatively, find out a way to expose kubelet metrics to the user?

timstclair replied:
Done.

yujuhong commented:
Thanks for writing the doc! LGTM

timstclair commented:
@k8s-bot unit test this

dchen1107 commented:
cc/ @timothysc

timstclair commented:
As I mentioned before this document is clearly incomplete, but my hope is that it can be a starting point for contributions as people perform more tests and further develop the tools.

timstclair commented:
@k8s-bot unit test this

tallclair committed:
Add node performance measuring guide

Add a development guide for measuring performance of node components.
The purpose of this guide is threefold:

1. Document the nuances of measuring kubelet performance so we don't
forget or need to reinvent the wheel.
2. Make it easier for new contributors to analyze performance.
3. Share tips and tricks that current team members might not be aware
of.

timstclair commented:
Squashed, rebased, and reran hack/update-generated-docs.sh.
